### PR TITLE
[Fix] dynamic switch for uncommon resolutions

### PR DESCRIPTION
**Pull request by lixiang007666**
This PR is done:

- [x] Fix https://github.com/siliconflow/onediff/issues/527 

Run:

```
python examples/text_to_image_sdxl_uncommon_res.py --run_multiple_resolutions true
```

### CODE REVIEW

File: src/infer_compiler_registry/register_diffusers/unet_2d_blocks_oflow.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+            # upsample_nearest_nhwc fails with large batch sizes. see https://github.com/huggingface/diffusers/issues/984
+            if hidden_states.shape[0] >= 64:
+                hidden_states = hidden_states.contiguous()
+
+            # if `output_size` is passed we force the interpolation output
+            # size and do not make use of `scale_factor=2`
+            if self.interpolate:
+                if output_size is None:
+                    hidden_states = F.interpolate(hidden_states, scale_factor=2.0, mode="nearest")
+                else:
```
Comment (lixiang007666) [inline src/infer_compiler_registry/register_diffusers/unet_2d_blocks_oflow.py:120]: 需要对 uncommon 的分辨率 warmup，才能保证后面碰到 uncommon 的分辨率正常切换。

Comment (hjchen2) [reply src/infer_compiler_registry/register_diffusers/unet_2d_blocks_oflow.py:120]: 可以只替换forward接口，Module不用整个替换，参考unet_2d_condition_oflow那个文件

File: src/infer_compiler_registry/register_diffusers/unet_2d_condition_oflow.py [hunk 2]
```diff
@@ -431,7 +431,9 @@ def forward(
             # if we have not reached the final block and need to forward the
             # upsample size, we do it here
             if not is_final_block and forward_upsample_size:
-                upsample_size = down_block_res_samples[-1].shape[2:]
+                # To support dynamic switching of special resolutions, pass a like tensor.
```
Comment (hjchen2) [inline src/infer_compiler_registry/register_diffusers/unet_2d_condition_oflow.py:434]: 可以尝试强制把forward_upsample_size打开，这样任意分辨率都能走到interpolate_like那个分支了

File: src/infer_compiler_registry/register_diffusers/unet_2d_blocks_oflow.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+                hidden_states = hidden_states.contiguous()
+
+            # if `output_size` is passed we force the interpolation output
+            # size and do not make use of `scale_factor=2`
+            if output_size is None:
+                hidden_states = F.interpolate(hidden_states, scale_factor=2.0, mode="nearest")
+            else:
+                # Rewritten for the switching of uncommon resolutions.
+                # hidden_states = F.interpolate(hidden_states, size=output_size, mode="nearest")
+                hidden_states = F.interpolate_like(hidden_states, like=output_size, mode="nearest")
```
Comment (hjchen2) [inline src/infer_compiler_registry/register_diffusers/unet_2d_blocks_oflow.py:226]: 我记得interpolate_like只支持了2d，这里最好和comfy那里对齐一下

File: examples/text_to_image_sdxl_uncommon_res.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    "--base", type=str, default="stabilityai/stable-diffusion-xl-base-1.0"
+)
+parser.add_argument("--variant", type=str, default="fp16")
+parser.add_argument(
+    "--prompt",
+    type=str,
+    default="street style, detailed, raw photo, woman, face, shot on CineStill 800T",
+)
+parser.add_argument("--height", type=int, default=512)
+parser.add_argument("--width", type=int, default=512)
```
Comment (hjchen2) [inline examples/text_to_image_sdxl_uncommon_res.py:25]: 这个文件是不是可以和sdxl多分辨率的单测合并一下

File: src/infer_compiler_registry/register_diffusers/unet_2d_blocks_oflow.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+LoRACompatibleConv = (
+    transformed_diffusers.models.lora.LoRACompatibleConv
+)
+
+try:
+    USE_PEFT_BACKEND = transformed_diffusers.utils.USE_PEFT_BACKEND
+except Exception as e:
+    USE_PEFT_BACKEND = False
+
+if diffusers_version >= version.parse("0.25.00"):
```
Comment (hjchen2) [inline src/infer_compiler_registry/register_diffusers/unet_2d_blocks_oflow.py:23]: 在forward函数内部做判断，减少冗余代码

**Review by hjchen2 — APPROVED** [review]

**Comment by lixiang007666** [comment]
我在本地用社区版的 oneflow 没复现出来 ci 的报错：https://github.com/siliconflow/onediff/actions/runs/7669755215/job/20904408049?pr=573
区别是我本地是 cu121，而 ci 是 cu118，后面换下版本再试一下。
